importers/aws/aws.go

```diff
@@ -0,0 +1,704 @@
+			if !ok {
+				instanceMap[*instance.InstanceId] = *i.sourceImage.ImageId
+
+				if err := setInstanceTag(*instance.InstanceId, "InUse", "true"); err != nil {
+					glog.Errorf("Error setting tag for instance %s: %v", *instance.InstanceId, err)
+				}
+
+				return instance, nil
+			}
+		}
+	}
+
+	return types.Instance{}, fmt.Errorf("unable to find available instance")
+}
+
+// region returns AWS availability region for the project.
+func (i *image) region() (string, error) {
+	dazOut, err := ec2Client.DescribeAvailabilityZones(context.TODO(), &ec2.DescribeAvailabilityZonesInput{})
+	if err != nil {
+		return "", err
+	}
+
+	var regionName string
+
+	for _, zone := range dazOut.AvailabilityZones {
+		if zone.RegionName != nil {
+			regionName = *zone.RegionName
+			return regionName, nil
+		}
+	}
+
+	return "", fmt.Errorf("error getting region name")
+}
+
+// copy copies Amazon owned AMI to HashR project.
+//
+// Note: This operation will create a new AMI on HashR project.
+// Disk image is created based on the snapshot in AMI in the HashR project.
+func (i *image) copy() error {
+	targetImageName := fmt.Sprintf("copy-%s", *i.sourceImage.ImageId)
+	ciout, err := ec2Client.CopyImage(context.TODO(), &ec2.CopyImageInput{
+		Name:          aws.String(targetImageName),
+		SourceImageId: aws.String(*i.sourceImage.ImageId),
+		SourceRegion:  aws.String(i.regionName),
+	})
+	if err != nil {
+		return err
+	}
+
+	glog.Infof("Copying source image  %s to HashR project image %s", *i.sourceImage.ImageId, *ciout.ImageId)
+
+	for w := 0; w < buildTimeout/100; w++ {
+		time.Sleep(30 * time.Second)
+
+		diout, err := ec2Client.DescribeImages(context.TODO(), &ec2.DescribeImagesInput{
+			ImageIds: []string{*ciout.ImageId},
+		})
+		if err != nil {
+			glog.Errorf("Error getting %s image details. %v", *ciout.ImageId, err)
+			return err
+		}
+
+		i.localImage = diout.Images[0]
+
+		if diout.Images[0].State == types.ImageStateAvailable {
+			break
+		}
+	}
+
+	if *i.localImage.ImageId == "" {
+		return fmt.Errorf("unable to get image Id")
+	}
+	return nil
+}
+
+// export exports a tar.gz disk image to AWS S3 bucket.
+func (i *image) export() error {
+	if len(i.localImage.BlockDeviceMappings) < 1 {
+		return fmt.Errorf("no block device for %s", *i.localImage.ImageId)
+	}
+
+	snapshotId := *i.localImage.BlockDeviceMappings[0].Ebs.SnapshotId
+	volumeSize := int32(*i.localImage.BlockDeviceMappings[0].Ebs.VolumeSize)
+	glog.Infof("Using snapshot %s of the image %s to create volume", snapshotId, *i.sourceImage.ImageId)
+
+	dsout, err := ec2Client.DescribeSnapshots(context.TODO(), &ec2.DescribeSnapshotsInput{
+		Filters: []types.Filter{
+			{
+				Name:   aws.String("snapshot-id"),
+				Values: []string{snapshotId},
+			},
+		},
+	})
+	if err != nil {
+		glog.Errorf("Error getting details of snapshot %s: %v", snapshotId, err)
+		return nil
+	}
+	snapshot := dsout.Snapshots[0]
+
+	i.volumeId = *snapshot.VolumeId
+	if i.volumeId == "vol-ffffffff" {
```

Comment:
`vol-ffffffff` indicates no volume exists for a snapshot. If an existing volume exists for the snapshot, the code uses that volume instead of creating a new volume.